addons/hr_holidays/models/resource.py

```diff
@@ -1,10 +1,136 @@
+            })
+            return leave.copy(copys)
+        if gto_start > leave_start\
+                and gto_start < leave_end\
+                and gto_end >= leave_end:
+            leave.write({
+                'date_to': leave_tz.localize(gto_start - timedelta(seconds=1))\
+                        .astimezone(UTC).replace(tzinfo=None)
+            })
+            return self.env['hr.leave'].sudo()
+
+    def _split_leave(self, leave, time_domain_dict):
+        new_leaves = self.env['hr.leave'].sudo()
+        for record in sorted(
+                filter(lambda r: r['date_to'] > leave['date_from'] and r['date_from'] < leave['date_to'], time_domain_dict),
+                key=lambda r: r['date_from']):
+            new_leave = self._split_leave_on_gto(leave, record)
+            if new_leave:
+                new_leaves |= new_leave
+        return new_leaves
+
+    def _reevaluate_leaves(self, time_domain_dict):
+        if time_domain_dict:
+            domain = self._get_domain(time_domain_dict)
+            leaves = self.env['hr.leave'].search(domain)
+            if not leaves:
+                return
+
+            previous_durations = [leave.number_of_days for leave in leaves]
+            previous_states = [leave.state for leave in leaves]
+            leaves.sudo().write({
+                'state': 'draft',
+            })
+            leaves._compute_number_of_days()
+            for previous_duration, leave, state in zip(previous_durations, leaves, previous_states):
+                duration_difference = previous_duration - leave.number_of_days
+                if duration_difference > 0 and leave['holiday_allocation_id']:
+                    leave.message_post(body=_("Due to a change in global time offs, you have been granted %s day(s) back", duration_difference), subtype_id=1)
+                if leave.number_of_days > previous_duration\
+                        and leave['holiday_status_id']['name'] not in ['Sick Time Off']:
+                    new_leaves = self._split_leave(leave, time_domain_dict)
+                    leaves |= new_leaves
+                    previous_states += [state for i in range(len(new_leaves))]
+
+            for index, leave in enumerate(leaves):
+                leave.write({'state': previous_states[index]})
+
+            for leave in leaves:
+                if leave.number_of_days == 0.0:
+                    leave.force_cancel(_("a new public holiday completely overrides this leave"), 1)
+        return True
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        res = super().create(vals_list)
+        time_domain_dict = res._get_time_domain_dict()
+        if time_domain_dict:
+            self._reevaluate_leaves(time_domain_dict)
+        return res
+
+    def write(self, vals):
+        time_domain_dict = self._get_time_domain_dict()
+        res = super().write(vals)
+        time_domain_dict = self._get_time_domain_dict(time_domain_dict)
```

Comment:
I feel like making `_get_time_domain_dict` take no argument and using extend is cleaner
```suggestion
        time_domain_dict.extend(self._get_time_domain_dict(time_domain_dict))
```